File: generator/on-board/src/main/scala/ru/mipt/acsl/geotarget/OnBoardCSourceGenerator.scala
package ru.mipt.acsl.geotarget

import java.io.File
import java.nio.charset.StandardCharsets.UTF_8

import com.google.common.io.Resources

import com.typesafe.scalalogging.LazyLogging
import ru.mipt.acsl.decode.c.generator.{CGeneratorConfiguration, CSourceGenerator, FileGeneratorConfiguration, GeneratorSource}
import ru.mipt.acsl.decode.model.naming.Fqn
import ru.mipt.acsl.decode.model.naming.Fqn
import ru.mipt.acsl.decode.parser.ModelRegistry

import scala.collection.immutable.HashMap
import scala.io.Source

/**
  * @author Artem Shein
  */
object OnBoardCSourceGenerator extends LazyLogging {

  val RootComponentFqn = "ru.mipt.acsl.photon.Main"

  def fqn(str: String): Fqn = Fqn(str)

  def main(args : Array[String]) = {
    val config = new CGeneratorConfiguration(new File("gen/"),
      OnBoardModelRegistry.registry,
      RootComponentFqn,
      HashMap(
        fqn("decode") -> Some(fqn("photon.decode")),
        fqn("ru.mipt.acsl.photon") -> Some(fqn("photon")),
        fqn("ru.mipt.acsl.foundation") -> Some(fqn("photon.foundation")),
        fqn("ru.mipt.acsl.fs") -> Some(fqn("photon.fs")),
        fqn("ru.mipt.acsl.identification") -> Some(fqn("photon.identification")),
        fqn("ru.mipt.acsl.mcc") -> Some(fqn("photon")),
        fqn("ru.mipt.acsl.routing") -> Some(fqn("photon.routing")),
        fqn("ru.mipt.acsl.scripting") -> Some(fqn("photon.scripting")),
        fqn("ru.mipt.acsl.segmentation") -> Some(fqn("photon.segmentation")),
        fqn("ru.mipt.acsl.tm") -> Some(fqn("photon.tm"))),
      sources = OnBoardModelRegistry.Sources.map(source =>
        GeneratorSource(ModelRegistry.sourceName(source), ModelRegistry.sourceContents(source))),
      isSingleton = true,
      includeModelInfo = true,
      prologue = FileGeneratorConfiguration(isActive = true, path = Some("photon/prologue.h"),
        contents = Some(
          """
            |#ifndef __PHOTON_PROLOGUE__
            |#define __PHOTON_PROLOGUE__
            |
            |#include "photon/Result.h"
            |#include "photon/Reader.h"
            |#include "photon/Writer.h"
            |#include "photon/Varuint.h"
            |
            |typedef uint8_t PhotonGtB8;
            |typedef uint8_t PhotonGtU8;
            |typedef uint16_t PhotonGtU16;
            |typedef uint32_t PhotonGtU32;
            |typedef uint64_t PhotonGtU64;
            |typedef PhotonBer PhotonGtBer;
            |
            |typedef struct {
            |   void* _stub;
            |} PhotonGcMain;
            |
            |#endif
          """.stripMargin)))
    logger.debug(s"Generating on-board sources to ${config.outputDir.getAbsolutePath}...")
    new CSourceGenerator(config).generate()
  }
}

File: generator/on-board/src/main/java/ru/mipt/acsl/geotarget/OnBoardCSourceGenerator.java
package ru.mipt.acsl.geotarget;

import com.google.common.io.Resources;
import ru.mipt.acsl.decode.c.generator.CGeneratorConfiguration;
import ru.mipt.acsl.decode.c.generator.CSourceGenerator;
import ru.mipt.acsl.decode.c.generator.FileGeneratorConfiguration;
import ru.mipt.acsl.decode.c.generator.GeneratorSource;
import ru.mipt.acsl.decode.model.naming.Fqn;
import ru.mipt.acsl.decode.parser.ModelRegistry;

import java.io.File;
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.util.HashMap;
import java.util.stream.Collectors;

/**
 * @author Artem Shein
 */
public class OnBoardCSourceGenerator {

    public static final String ROOT_COMPONENT_FQN_STRING = "mcc.photon.Main";
    public static final String PROLOGUE_PATH = "photon/prologue.h";

    public static Fqn fqn(String str) {
        return Fqn.newInstance(str);
    }

    public static void main(String[] args) {
        try {
            CGeneratorConfiguration config = CGeneratorConfiguration.newInstance(new File("gen/"),
                    OnBoardModelRegistry.registry(),
                    ROOT_COMPONENT_FQN_STRING,
                    new HashMap<Fqn, Fqn>() {{
                        put(fqn("decode"), fqn("photon.decode"));
                        put(fqn("mcc.photon"), fqn("photon"));
                        put(fqn("mcc.foundation"), fqn("photon.foundation"));
                        put(fqn("mcc.fs"), fqn("photon.fs"));
                        put(fqn("mcc.identification"), fqn("photon.identification"));
                        put(fqn("mcc.mcc"), fqn("photon"));
                        put(fqn("mcc.routing"), fqn("photon.routing"));
                        put(fqn("mcc.scripting"), fqn("photon.scripting"));
                        put(fqn("mcc.segmentation"), fqn("photon.segmentation"));
                        put(fqn("mcc.tm"), fqn("photon.tm"));
                    }},
                    OnBoardModelRegistry.Sources.ALL.stream()
                            .map(source -> GeneratorSource.newInstance(ModelRegistry.sourceName(source),
                                    ModelRegistry.sourceContents(source))).collect(Collectors.toList()),
                    FileGeneratorConfiguration.newInstance(true, PROLOGUE_PATH,
                            Resources.toString(Resources.getResource(PROLOGUE_PATH), StandardCharsets.UTF_8)),
                    FileGeneratorConfiguration.newInstance(),
                    true, true);
            System.out.println(String.format("Generating on-board sources to %s...",
                    config.outputDir().getAbsolutePath()));
            new CSourceGenerator(config).generate();
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

}